operations.go

```diff
@@ -0,0 +1,67 @@
+package webrtc
+
+import (
+	"sync"
+)
+
+// Operation is a function
+type operation func()
+
+// Operations is a task executor.
+type operations struct {
+	ops     []operation
+	mu      sync.Mutex
+	startMu sync.Mutex
+	done    chan struct{}
+}
+
+// Enqueue adds a new action to be executed. If there are no actions scheduled,
+// the execution will start immediately in a new goroutine.
+func (o *operations) Enqueue(op operation) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.ops = append(o.ops, op)
+	if len(o.ops) == 1 {
+		done := make(chan struct{})
+		o.done = done
+
+		go func() {
+			o.startMu.Lock()
+			defer o.startMu.Unlock()
+			o.start()
+			close(done)
+		}()
+	}
+}
+
+// Done will return a channel that will be closed as soon as all currently
+// enqueued operations are finished.
+func (o *operations) Done() <-chan struct{} {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.done != nil {
+		return o.done
+	}
+	// no queued operations, return the closed channel so reading from it  does not block
+	done := make(chan struct{})
+	close(done)
+	return done
+}
+
+func (o *operations) start() {
+	for {
```

Comment:
What do you think of having a local function that does the popping? It can hold the lock as well. This code is good now, but I am afraid a bug fix/modification could break it in the future :(

```
popOperation := func() func() {
    o.mu.Lock()
    defer o.mu.Unlock()
    
    if len(o.ops) == 0 {
        return nil
    }
 
   fn := o.ops[0]
   o.ops = o.ops[1:]
   return fn
}

```

Reply:
You're right. I didn't like doing the unlock on two different places. I came up with something a little simpler based on your other comment:

```
func (o *operations) pop() (fn func(), isLast bool) {
	o.mu.Lock()
	defer o.mu.Unlock()
	fn = o.ops[0]
	o.ops = o.ops[1:]
	return fn, len(o.ops) == 0
}
```